doc/sphinx/language/core/records.rst

```diff
@@ -3,208 +3,313 @@
 
-The :term:`command` generates also, when it is possible, the projection
-functions for destructuring an object of type :token:`ident`. These
-projection functions are given the names of the corresponding
-fields. If a field is named `_` then no projection is built
-for it. In our example:
+         (* Application form: use the constructor and provide values for all the fields
+            in order.  "mkRat" is defined by the Record command *)
+         Definition half' := mkRat true 1 2 (O_S 1) one_two_irred.
 
-.. coqtop:: all
+Accessing fields (projections)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-  Eval compute in top half.
-  Eval compute in bottom half.
-  Eval compute in Rat_bottom_cond half.
+   .. insertprodn term_projection term_projection
 
-An alternative syntax for projections based on a dot notation is
-available:
+   .. prodn::
+      term_projection ::= @term0 .( @qualid {? @univ_annot } {* @arg } )
+      | @term0 .( @ @qualid {? @univ_annot } {* @term1 } )
 
-.. coqtop:: all
+   The value of a field can be accessed using *projection form* (:n:`@term_projection`,
+   shown here) or with *application form* (see :n:`@term_application`) using the
+   projection function associated with the field.  Don't forget the parentheses for the
+   projection form.
+   Glossing over some syntactic details, the two forms are:
 
-   Eval compute in half.(top).
+   - :n:`@qualid__record.( {? @ } @qualid__field {* @arg })` (projection) and
 
-.. flag:: Printing Projections
+   - :n:`{? @ } @qualid__field {* @arg } @qualid__record` (application)
 
-   This :term:`flag` activates the dot notation for printing.
+   where the :n:`@arg`\s are the parameters of the inductive type.  If :n:`@` is
+   specified, all implicit arguments must be provided.
 
-   .. example::
+   Since the projected object is part of the notation, it is always
+   considered an explicit argument of :token:`qualid`, even if it is
+   formally declared as implicit (see :ref:`ImplicitArguments`).
+
+   .. exn:: @ident cannot be defined because it is informative and @ident is not
+
+      For example, :n:`Record foo:Prop := { x:Type }` generates the message
+      "x cannot be defined ... and foo is not".  Proofs (objects of sort :n:`Prop`)
+      are supposed to be non-distinguishable.  If you have two inhabitants of
+      :n:`Type`, such as :n:`%{%| x := nat %|%}` and :n:`%{%| x := bool %|%}`, they are
+      distinguishable (i.e. informative) and are therefore prohibited.
+
+   .. example:: Accessing record fields
 
       .. coqtop:: all
 
-         Set Printing Projections.
-         Check top half.
+         (* projection form *)
+         Eval compute in half.(top).
+
+      .. coqtop:: in
+
+         Goal True.
+
+      .. coqtop:: all
+
+         let x := eval compute in half.(top) in idtac x.
+
+         (* application form *)
+         Eval compute in top half.
+         Eval compute in bottom half.
+         Eval compute in Rat_bottom_nonzero half.
```

Comment:
What is the point of showing three examples in application form but only one in projection form?